Bump up minimum C++ standard to C++14

maxpkatz opened the pull request:
## PR summary

C++14 is now required as the minimum C++ standard. This should cause no changes except for RHEL 7/CentOS 7 systems where users will be required to use a more recent gcc than the system default.

## PR motivation

This will support some C++14 constructs we want to add to Microphysics.

## PR checklist

- [x] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite to more than roundoff level
- [ ] all newly-added functions have docstrings as per the coding conventions
- [x] the `CHANGES` file has been updated, if appropriate
- [x] if appropriate, this change is described in the docs

zingale commented:
tests pass: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2020-10-31/index.html

maxpkatz commented:
Since this would break the case where we use COMP=PGI on Summit (which blocks us from loading a new enough gcc module), I added a workaround to hardcode the gcc used by nvcc in Make.Castro. This workaround can be removed when we have switched over all of the problem setups to not require PGI anymore.
